### PR TITLE
Add header matched routing

### PR DESCRIPTION
**Pull request by junr03**

### CODE REVIEW

File: source/common/router/config_impl.h [hunk 1]
```diff
@@ -211,6 +211,7 @@ class RouteEntryImplBase : public RouteEntry, public Matchable, public RedirectE
   const RateLimitPolicyImpl rate_limit_policy_;
   const ShadowPolicyImpl shadow_policy_;
   const Upstream::ResourcePriority priority_;
+  std::vector<Json::Object> config_headers_;
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: No JSON parsing in hot path. Unpack at config load time to make sure everything is correct, and store in struct.

File: source/common/router/config_impl.cc [hunk 2]
```diff
@@ -79,6 +83,15 @@ bool RouteEntryImplBase::matches(const Http::HeaderMap& headers, uint64_t random
     matches &= (headers.get(Http::Headers::get().ContentType) == content_type_);
   }
 
+  if (!config_headers_.empty()) {
+    for (const Json::Object& header_map : config_headers_) {
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: i'd also include contentType header here (and do not have separate content_type_) unless i'm missing something

Comment (mattklein123) [inline source/common/router/config_impl.cc]: That is a breaking change. We can look into that separately (open internal JIRA), but let's not do that as part of this change.

File: source/common/router/config_impl.cc [hunk 3]
```diff
@@ -84,8 +98,8 @@ bool RouteEntryImplBase::matches(const Http::HeaderMap& headers, uint64_t random
   }
 
   if (!config_headers_.empty()) {
-    for (const Json::Object& header_map : config_headers_) {
-      matches &= headers.get(Http::LowerCaseString(header_map.getString("header_name"))) == header_map.getString("header_value");
+    for (const HeaderData header_data : config_headers_) {
```
Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: const HeaderData&

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -79,6 +97,15 @@ bool RouteEntryImplBase::matches(const Http::HeaderMap& headers, uint64_t random
     matches &= (headers.get(Http::Headers::get().ContentType) == content_type_);
   }
 
+  if (!config_headers_.empty()) {
+    for (const HeaderData header_data : config_headers_) {
+      matches &= headers.get(Http::LowerCaseString(header_data.header_name_)) == header_data.header_value_;
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: store as lower case string in the struct

File: source/common/router/config_impl.cc [hunk 5]
```diff
@@ -65,6 +65,24 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHost& vhost, const Json::Obj
...
+      HeaderData header_data;
+
+      if (!header_map.hasObject("header_name")) {
+        throw EnvoyException("headers in the route config must have a header name");
+      } else {
+        header_data.header_name_ = header_map.getString("header_name");
+      }
+
+      // allow header value to be empty, allows matching to be only based on header presence.
+      header_data.header_value_ = header_map.getString("header_value");
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: needs default value of ""

File: source/common/router/config_impl.cc [hunk 6]
```diff
@@ -65,6 +65,24 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHost& vhost, const Json::Obj
...
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
+
+  if (route.hasObject("headers")) {
+    std::vector<Json::Object> config_headers = route.getObjectArray("headers");
+    for (const Json::Object& header_map : config_headers) {
+      HeaderData header_data;
+
+      if (!header_map.hasObject("header_name")) {
+        throw EnvoyException("headers in the route config must have a header name");
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: don't bother with this check here. Just let JsonException be thrown if value does not exist

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: storing as Http::LowerCaseString

File: source/common/router/config_impl.cc (hunk 5 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: added.

File: source/common/router/config_impl.cc (hunk 6 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: will let JsonException be thrown.

File: source/common/router/config_impl.h [hunk 7]
```diff
@@ -194,6 +194,12 @@ class RouteEntryImplBase : public RouteEntry, public Matchable, public RedirectE
     uint64_t default_;
   };
 
+  struct HeaderData {
+    HeaderData(const std::string& header_name, const std::string& header_value);
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: constructor not needed

File: source/common/router/config_impl.h [hunk 8]
```diff
@@ -194,6 +194,12 @@ class RouteEntryImplBase : public RouteEntry, public Matchable, public RedirectE
...
+    Http::LowerCaseString header_name_;
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: const

File: source/common/router/config_impl.h [hunk 9]
```diff
@@ -194,6 +194,12 @@ class RouteEntryImplBase : public RouteEntry, public Matchable, public RedirectE
...
+    std::string header_value_;
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: const

File: source/common/router/config_impl.cc [hunk 10]
```diff
@@ -65,6 +65,16 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHost& vhost, const Json::Obj
...
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
+
+  if (route.hasObject("headers")) {
+    std::vector<Json::Object> config_headers = route.getObjectArray("headers");
+    for (const Json::Object& header_map : config_headers) {
+      // allow header value to be empty, allows matching to be only based on header presence.
+      HeaderData header_data(header_map.getString("header_name"),
+                             header_map.getString("header_value", ""));
+      config_headers_.push_back(header_data);
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: push_back struct directly

Comment (RomanDzhabarov) [inline source/common/router/config_impl.cc]: you can emplace_back(header_map.getString("header_name"), header_map.getString("header_value", ""))

File: source/common/router/config_impl.h (hunk 7 above)
Comment (junr03) [inline source/common/router/config_impl.h]: deleted

File: source/common/router/config_impl.h (hunk 8 above)
Comment (junr03) [inline source/common/router/config_impl.h]: fixed

File: source/common/router/config_impl.cc (hunk 10 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: looks like emplace_back would not allow me to do [uniform initialization](http://stackoverflow.com/questions/10890653/why-would-i-ever-use-push-back-instead-of-emplace-back)

Comment (junr03) [inline source/common/router/config_impl.cc]: done

File: source/common/router/config_impl.h (hunk 9 above)
Comment (junr03) [inline source/common/router/config_impl.h]: same question

File: source/common/router/config_impl.h (hunk 1 above)
Comment (junr03) [inline source/common/router/config_impl.h]: done

File: source/common/router/config_impl.cc (hunk 3 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: done

File: source/common/router/config_impl.cc [hunk 11]
```diff
@@ -65,6 +65,15 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHost& vhost, const Json::Obj
   if (!(isRedirect() ^ !cluster_name_.empty())) {
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
+
+  if (route.hasObject("headers")) {
+    std::vector<Json::Object> config_headers = route.getObjectArray("headers");
+    for (const Json::Object& header_map : config_headers) {
+      // allow header value to be empty, allows matching to be only based on header presence.
+      config_headers_.emplace_back(Http::LowerCaseString(header_map.getString("header_name")),
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: "name"

File: source/common/router/config_impl.cc [hunk 12]
```diff
@@ -65,6 +65,15 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHost& vhost, const Json::Obj
...
+                                   header_map.getString("header_value", EMPTY_STRING));
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: "value"

File: source/common/router/config_impl.cc (hunk 11 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: changed.

File: docs/configuration/http_conn_man/route_config/route.rst [hunk 13]
```diff
@@ -23,7 +23,8 @@ next (e.g., redirect, forward, rewrite, etc.).
     "retry_policy": "{...}",
     "rate_limit": "{...}",
     "shadow": "{...}",
-    "priority": "..."
+    "priority": "...",
+    "headers": "[...]"
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route.rst]: "headers": []

(look at other docs to match formatting)

File: docs/configuration/http_conn_man/route_config/route.rst [hunk 14]
```diff
@@ -94,6 +95,9 @@ priority
   *(optional, string)* Optionally specifies the :ref:`routing priority
   <arch_overview_http_routing_priority>`.
 
+:ref:`headers <config_http_conn_man_route_table_route_headers>`
+  *(optional, array)* Allows to do header matched routing with arbitrary headers.
```
Comment (mattklein123) [inline docs/configuration/http_conn_man/route_config/route.rst]: grammar

File: source/common/router/config_impl.cc [hunk 15]
```diff
@@ -79,6 +88,19 @@ bool RouteEntryImplBase::matches(const Http::HeaderMap& headers, uint64_t random
     matches &= (headers.get(Http::Headers::get().ContentType) == content_type_);
   }
 
+  if (!config_headers_.empty()) {
+    for (const HeaderData& header_data : config_headers_) {
+      if (header_data.value_ == EMPTY_STRING) {
+        matches &= headers.has(header_data.name_);
+      } else {
+        matches &= headers.get(header_data.name_) == header_data.value_;
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: nit: parens around boolean expression

File: source/common/router/config_impl.h [hunk 16]
```diff
@@ -194,6 +194,14 @@ class RouteEntryImplBase : public RouteEntry, public Matchable, public RedirectE
     uint64_t default_;
   };
 
+  struct HeaderData {
+    HeaderData(const Http::LowerCaseString& name, const std::string& value)
+        : name_(name), value_(value){};
```
Comment (mattklein123) [inline source/common/router/config_impl.h]: nit: remove trailing ;

File: docs/configuration/http_conn_man/route_config/route.rst (hunk 13 above)
Comment (junr03) [inline docs/configuration/http_conn_man/route_config/route.rst]: fixed

File: docs/configuration/http_conn_man/route_config/route.rst (hunk 14 above)
Comment (junr03) [inline docs/configuration/http_conn_man/route_config/route.rst]: new sentence.

File: source/common/router/config_impl.cc (hunk 15 above)
Comment (junr03) [inline source/common/router/config_impl.cc]: done

**Pull request merged**